Add: [Script] Unbunch order flag

SamuXarick opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem
Missing unbunch feature for AIs.
<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description
Allow AIs to unbunch vehicles.
Proper handle the new `OF_UNBUNCH_IN_DEPOT` order flag when combined with other existing order flags. Add regression tests to test this flag and the returned error strings associated with it.
<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations
Only for API version 15?
<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

SamuXarick committed:
Add: [Script] Unbunch order flag

Allow AIs to unbunch vehicles.
Proper handle the new OF_UNBUNCH_IN_DEPOT order flag when combined with other existing order flags.
Add regression tests to test this flag and the returned error strings associated with it.

SamuXarick committed:
Change: [Script] Use only one unbunch error string instead of five

SamuXarick commented on src/script/api/script_order.cpp:
```diff
@@ -515,18 +519,18 @@ static ScriptOrder::OrderPosition RealOrderPositionToScriptOrderPosition(Vehicle
 
 		case OT_GOTO_WAYPOINT:
 			order.MakeGoToWaypoint(::GetStationIndex(destination));
+			if (v->type == VEH_TRAIN) order.SetNonStopType((OrderNonStopFlags)GB(order_flags, 0, 2));
```
```suggestion
			if (v->IsGroundVehicle()) order.SetNonStopType((OrderNonStopFlags)GB(order_flags, 0, 2));
```
This probably should be changed in the future when road waypoints are properly supported (#13398, #13419 and maybe some others...)

LordAro commented on src/order_cmd.cpp:
```diff
@@ -799,10 +799,11 @@ CommandCost CmdInsertOrder(DoCommandFlags flags, VehicleID veh, VehicleOrderID s
+				if (new_order.GetDepotActionType() & ODATFB_HALT) return CMD_ERROR;
+				if (v->HasFullLoadOrder()) return CommandCost(STR_ERROR_UNBUNCHING_NO_UNBUNCHING_FULL_LOAD);
+				if (v->HasUnbunchingOrder()) return CommandCost(STR_ERROR_UNBUNCHING_ONLY_ONE_ALLOWED);
+				if (v->HasConditionalOrder()) return CommandCost(STR_ERROR_UNBUNCHING_NO_UNBUNCHING_CONDITIONAL);
```
i'm not clear what this change is doing. adding a script function shouldn't be making any changes to the "actual" code

SamuXarick replied:
These errors are only reachable via AIs, human players won't see it.
AIs however don't have the extra error message implemented in them, and the error translation from the original code to AI would only be taking the first message into account, `STR_ERROR_CAN_T_ADD_ORDER`, which isn't clear enough it's related to unbunch.

LordAro replied:
I don't see any evidence this is AI only code, and you're removing the first part of the error message that would be displayed, no?

SamuXarick replied:
Oh, and the other one is an invalid state which only AIs have the ability to set, but shouldn't. Setting both unbunch + stop at depot is supposed to be not possible.

SamuXarick replied:
With 2 error messages:
```
GetLastErrorString(): ERR_UNKNOWN
GetErrorCategory(): 0 // ERR_CAT_NONE
```

with 1 error message:
```
GetLastErrorString(): ERR_ORDER_UNBUNCH
GetErrorCategory(): 10 // ERR_CAT_ORDER
```

If the 2 error messages are to be kept, I can map `STR_ERROR_CAN_T_ADD_ORDER` to script `ERR_CAN_T_ADD_ORDER`, and the result would be:
```
GetLastErrorString(): ERR_CAN_T_ADD_ORDER
GetErrorCategory(): 10 // ERR_CAT_ORDER
```
But this wouldn't exactly tell it's related to unbunch. It could be related to something else, like... incorrect station type for vehicle.

Rito13 replied:
These strings are used in 3 places:
1. https://github.com/OpenTTD/OpenTTD/blob/2d8cb5314c34af6c8f5be57868b3c8b6a2192110/src/order_gui.cpp#L382
Where user has to have controll pressed.
2. https://github.com/OpenTTD/OpenTTD/blob/2d8cb5314c34af6c8f5be57868b3c8b6a2192110/src/order_cmd.cpp#L1278
Which is shown when user modifies the orders.
3. The one you have changed, it seems like when new order will be added without use of ctrl you change will be visible to the player.

SamuXarick replied:
In order_gui.cpp, `GetOrderCmdFromTile` has its own checks for adding an unbunch order, which protect it from the invalid condition checks in order_cmd.cpp's `CmdInsertOrder`.

If Ctrl isn't pressed, it’s not considered an unbunch order, and `new_order.GetDepotActionType().Test(OrderDepotActionFlag::Unbunch)` in `CmdInsertOrder` prevents it from ever reaching the error message checks.

From a player's perspective, `GetOrderCmdFromTile` makes it unlikely to reach `CmdInsertOrder` in a way that would trigger its errors to be shown. Instead, those errors would originate from `order_gui.cpp` in the `OnPlaceObject` function:
https://github.com/OpenTTD/OpenTTD/blob/2d8cb5314c34af6c8f5be57868b3c8b6a2192110/src/order_gui.cpp#L1471
The error it produces is only 2 lines, not 3.
<img width="273" height="62" alt="image" src="https://github.com/user-attachments/assets/952fb198-019a-4573-8dc8-dd66de9630ae" />

Hypothetically, if errors in the order_cmd `CmdInsertOrder` were to occur, they could look something like this:
<img width="276" height="77" alt="image" src="https://github.com/user-attachments/assets/0ae66ea6-256d-4068-a1f6-bc31afb6e5b1" />

Rito13 replied:
You are right in terms of singleplayer, the only way to insert new unbunch order is to select the depot with go to tool while pressing ctrl, then when the order can't be inserted the error message will occur from `GetOrderCmdFromTile` called by `OnPlaceObject` as you have said, but on multiplayer there is slight delay between the server and client. In very specific situation, when two players change orders of one vehicle at the same time, it can end up that when client adds an order everything will be fine, and error in `GetOrderCmdFromTile` will not occur, but cmd from another user could get to the server first, then on the server `CmdInsertOrder` will be called directly and will fail as the orders have changed after the last check on the client side, then server will send error message back to the client and it will be show.
I was able to reproduce that with my PC and mobile (on mobile I have only OTTD 14.1, so thats why I have used that version), the server was hasted from the phone, and screenshot was taken on the PC.

<img width="1888" height="1056" alt="Zrzut ekranu z 2026-01-13 14-03-33" src="https://github.com/user-attachments/assets/afbb924f-5058-4019-b990-87e0e6514da3" />

SamuXarick replied:
Oops, so it's possible to have the 3 lines message to show.

I'd still prefer to reduce it to 2 lines, to match the way the error comes from the gui side. The middle line infers incorrectly that it's a station order when we're adding a depot order. It would be more consistent and at the same time it allow the AI side of things to get the single error message, avoiding #12269 altogether. It's a simpler fix/implementation.

Rito13 replied:
I agree, the `STR_ERROR_CAN_T_ADD_ORDER` is just a wrong string to be used here, it should not be used with depots.
> STR_ERROR_CAN_T_ADD_ORDER                                       :{WHITE}... vehicle can't go to that station

SamuXarick committed:
Codechange: [Script] Reorder variable declarations for clarity

Introduce a local vehicle pointer variable to avoid repeated calls to ::Vehicle::Get.

SamuXarick committed:
Change: [Script] Silently ignore OrderNonStopFlags in ScriptOrder::InsertOrder for ships and aircraft

Non-stop order flags are only valid for road vehicles and trains. Silently ignore provided non-stop order flags for ships and aircraft.

SamuXarick committed:
Codechange: [Script] Set OrderDepotActionFlag::NearestDepot later to avoid repeated check
